File: microscope/python/alignment_microscope/sae.py
"""Sparse Autoencoder (SAE) integration module.

This module provides integration with Sparse Autoencoders for interpretability:

- SAEWrapper: Unified interface for SAE encoding/decoding
- SAEFeatures: Container for SAE feature activations
- SAEAnalyzer: Analysis tools for SAE features
- SAELens integration for loading pre-trained SAEs
"""

from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple, Union

import numpy as np


@dataclass
class SAEConfig:
    """Configuration for a Sparse Autoencoder."""

    d_in: int
    """Input dimension (model hidden size)."""

    d_sae: int
    """SAE feature dimension."""

    activation: str = "relu"
    """Activation function: 'relu', 'topk', 'jumprelu'."""

    k: Optional[int] = None
    """Top-k value for topk activation."""

    hook_point: str = ""
    """Hook point name (e.g., 'blocks.0.hook_resid_post')."""

    layer: int = 0
    """Layer index this SAE is associated with."""


class SAEFeatures:
    """Container for SAE feature activations.

    Provides utilities for analyzing which features are active,
    computing sparsity, and finding top-k features.
    """

    def __init__(
        self,
        activations: np.ndarray,
        config: Optional[SAEConfig] = None,
    ):
        """Initialize SAE features.

        Args:
            activations: Feature activations, shape [batch, seq_len, d_sae] or [positions, d_sae]
            config: Optional SAE configuration
        """
        self._activations = activations
        self._config = config
        self._compute_stats()

    def _compute_stats(self):
        """Compute sparsity statistics."""
        total = self._activations.size
        zeros = np.sum(self._activations == 0)
        self._sparsity = zeros / total if total > 0 else 0.0

        # Mean active features per position
        if self._activations.ndim >= 2:
            active_per_pos = np.sum(self._activations > 0, axis=-1)
            self._mean_active = float(np.mean(active_per_pos))
        else:
            self._mean_active = float(np.sum(self._activations > 0))

    @property
    def activations(self) -> np.ndarray:
        """Get raw activations array."""
        return self._activations

    @property
    def shape(self) -> Tuple[int, ...]:
        """Get shape of activations."""
        return self._activations.shape

    @property
    def d_sae(self) -> int:
        """Get SAE feature dimension."""
        return self._activations.shape[-1]

    @property
    def sparsity(self) -> float:
        """Get sparsity ratio (fraction of zero activations)."""
        return self._sparsity

    @property
    def mean_active_features(self) -> float:
        """Get mean number of active features per position."""
        return self._mean_active

    def active_features(self, threshold: float = 0.0) -> List[np.ndarray]:
        """Get indices of active features per position.

        Args:
            threshold: Minimum activation value to be considered active

        Returns:
            List of arrays, each containing active feature indices for a position
        """
        # Reshape to 2D: [positions, d_sae]
        flat = self._activations.reshape(-1, self.d_sae)
        return [np.where(row > threshold)[0] for row in flat]

    def top_k_features(self, k: int = 10) -> List[List[Tuple[int, float]]]:
        """Get top-k features per position with their activation values.

        Args:
            k: Number of top features to return

        Returns:
            List of lists of (feature_idx, activation_value) tuples
        """
        flat = self._activations.reshape(-1, self.d_sae)
        results = []

        for row in flat:
            # Get indices that would sort descending
            sorted_indices = np.argsort(row)[::-1][:k]
            top_k = [(int(idx), float(row[idx])) for idx in sorted_indices if row[idx] > 0]
            results.append(top_k)

        return results

    def feature_frequency(self, threshold: float = 0.0) -> np.ndarray:
        """Compute how often each feature activates across positions.

        Args:
            threshold: Minimum activation to count as active

        Returns:
            Array of shape [d_sae] with activation frequencies
        """
        flat = self._activations.reshape(-1, self.d_sae)
        num_positions = flat.shape[0]
        active_counts = np.sum(flat > threshold, axis=0)
        return active_counts / num_positions

    def to_dict(self) -> Dict[str, Any]:
        """Export to dictionary."""
        return {
            "activations": self._activations.tolist(),
            "shape": list(self.shape),
            "sparsity": self._sparsity,
            "mean_active_features": self._mean_active,
        }


class SAEWrapper:
    """Unified interface for Sparse Autoencoders.

    Supports loading SAEs from SAELens or custom weights.
    Provides encode/decode operations and feature analysis.
    """

    def __init__(
        self,
        w_enc: np.ndarray,
        w_dec: np.ndarray,
        b_enc: Optional[np.ndarray] = None,
        b_dec: Optional[np.ndarray] = None,
        config: Optional[SAEConfig] = None,
    ):
        """Initialize SAE wrapper with weights.

        Args:
            w_enc: Encoder weights [d_in, d_sae]
            w_dec: Decoder weights [d_sae, d_in]
            b_enc: Encoder bias [d_sae]
            b_dec: Decoder bias [d_in]
            config: SAE configuration
        """
        self.w_enc = w_enc.astype(np.float32)
        self.w_dec = w_dec.astype(np.float32)
        self.b_enc = b_enc.astype(np.float32) if b_enc is not None else None
        self.b_dec = b_dec.astype(np.float32) if b_dec is not None else None

        if config is None:
            config = SAEConfig(
                d_in=w_enc.shape[0],
                d_sae=w_enc.shape[1],
            )
        self.config = config

    @classmethod
    def from_saelens(
        cls,
        model_name: str,
        hook_point: str,
        layer: Optional[int] = None,
        device: str = "cpu",
    ) -> "SAEWrapper":
        """Load a pre-trained SAE from SAELens.

        Args:
            model_name: Model name in SAELens format (e.g., 'gpt2-small')
            hook_point: Hook point (e.g., 'blocks.0.hook_resid_post')
            layer: Optional layer index to extract from hook_point
            device: Device to load to ('cpu' or 'cuda')

        Returns:
            SAEWrapper instance

        Raises:
            ImportError: If SAELens is not installed
            ValueError: If SAE not found
        """
        try:
            from sae_lens import SAE
        except ImportError:
            raise ImportError(
                "SAELens is required for loading pre-trained SAEs. "
                "Install with: pip install sae-lens"
            )

        # Load SAE from SAELens
        sae = SAE.from_pretrained(
            release=model_name,
            sae_id=hook_point,
            device=device,
        )[0]  # Returns (sae, cfg, sparsity)

        # Extract weights
        w_enc = sae.W_enc.detach().cpu().numpy()
        w_dec = sae.W_dec.detach().cpu().numpy()
        b_enc = sae.b_enc.detach().cpu().numpy() if hasattr(sae, 'b_enc') else None
        b_dec = sae.b_dec.detach().cpu().numpy() if hasattr(sae, 'b_dec') else None

        # Determine layer from hook_point if not provided
        if layer is None and "blocks." in hook_point:
            layer = int(hook_point.split("blocks.")[1].split(".")[0])

        config = SAEConfig(
            d_in=w_enc.shape[0],
            d_sae=w_enc.shape[1],
            activation="relu",  # SAELens default
            hook_point=hook_point,
            layer=layer or 0,
        )

        return cls(w_enc, w_dec, b_enc, b_dec, config)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "SAEWrapper":
        """Load SAE from dictionary.

        Args:
            data: Dictionary with weights and config

        Returns:
            SAEWrapper instance
        """
        w_enc = np.array(data["w_enc"])
        w_dec = np.array(data["w_dec"])
        b_enc = np.array(data["b_enc"]) if "b_enc" in data else None
        b_dec = np.array(data["b_dec"]) if "b_dec" in data else None

        config = None
        if "config" in data:
            config = SAEConfig(**data["config"])

        return cls(w_enc, w_dec, b_enc, b_dec, config)

    @property
    def d_in(self) -> int:
        """Get input dimension."""
        return self.config.d_in

    @property
    def d_sae(self) -> int:
        """Get SAE feature dimension."""
        return self.config.d_sae

    def encode(self, activations: np.ndarray) -> SAEFeatures:
        """Encode activations to SAE features.

        Args:
            activations: Input activations [..., d_in]

        Returns:
            SAEFeatures container
        """
        original_shape = activations.shape
        # Flatten to 2D: [positions, d_in]
        flat = activations.reshape(-1, self.d_in).astype(np.float32)

        # Center by decoder bias
        if self.b_dec is not None:
            flat = flat - self.b_dec

        # Encode: x @ W_enc + b_enc
        features = flat @ self.w_enc
        if self.b_enc is not None:
            features = features + self.b_enc

        # Apply activation
        if self.config.activation == "relu":
            features = np.maximum(features, 0)
        elif self.config.activation == "topk" and self.config.k is not None:
            features = self._apply_topk(features, self.config.k)
        elif self.config.activation == "jumprelu":
            features = np.maximum(features, 0)  # Simplified JumpReLU

        # Reshape to match original batch/seq structure
        new_shape = list(original_shape[:-1]) + [self.d_sae]
        features = features.reshape(new_shape)

        return SAEFeatures(features, self.config)

    def _apply_topk(self, features: np.ndarray, k: int) -> np.ndarray:
        """Apply top-k sparsification.

        Args:
            features: Feature activations [positions, d_sae]
            k: Number of features to keep per position

        Returns:
            Sparsified features
        """
        result = np.zeros_like(features)

        for i in range(features.shape[0]):
            top_indices = np.argsort(features[i])[-k:]
            for idx in top_indices:
                if features[i, idx] > 0:
                    result[i, idx] = features[i, idx]

        return result

    def decode(self, features: Union[SAEFeatures, np.ndarray]) -> np.ndarray:
        """Decode SAE features back to activations.

        Args:
            features: SAE features [..., d_sae]

        Returns:
            Reconstructed activations [..., d_in]
        """
        if isinstance(features, SAEFeatures):
            feat_arr = features.activations
        else:
            feat_arr = features

        original_shape = feat_arr.shape
        flat = feat_arr.reshape(-1, self.d_sae).astype(np.float32)

        # Decode: features @ W_dec + b_dec
        reconstructed = flat @ self.w_dec
        if self.b_dec is not None:
            reconstructed = reconstructed + self.b_dec

        # Reshape
        new_shape = list(original_shape[:-1]) + [self.d_in]
        return reconstructed.reshape(new_shape)

    def reconstruction_error(
        self,
        activations: np.ndarray,
        features: Optional[SAEFeatures] = None,
    ) -> float:
        """Compute mean squared reconstruction error.

        Args:
            activations: Original activations
            features: Optional pre-computed features

        Returns:
            MSE reconstruction error
        """
        if features is None:
            features = self.encode(activations)

        reconstructed = self.decode(features)
        diff = activations - reconstructed
        return float(np.mean(diff ** 2))

    def to_dict(self) -> Dict[str, Any]:
        """Export to dictionary."""
        result = {
            "w_enc": self.w_enc.tolist(),
            "w_dec": self.w_dec.tolist(),
            "config": {
                "d_in": self.config.d_in,
                "d_sae": self.config.d_sae,
                "activation": self.config.activation,
                "k": self.config.k,
                "hook_point": self.config.hook_point,
                "layer": self.config.layer,
            },
        }
        if self.b_enc is not None:
            result["b_enc"] = self.b_enc.tolist()
        if self.b_dec is not None:
            result["b_dec"] = self.b_dec.tolist()
        return result


class SAEAnalyzer:
    """Analyzer for SAE features across traces.

    Provides tools for:
    - Feature activation analysis
    - Cross-layer feature tracking
    - Behavior-specific feature identification
    """

    def __init__(self):
        """Initialize analyzer."""
        self._saes: Dict[str, SAEWrapper] = {}

    def register_sae(self, name: str, sae: SAEWrapper):
        """Register an SAE for analysis.

        Args:
            name: Unique name for this SAE (e.g., 'layer_0_resid')
            sae: SAE wrapper instance
        """
        self._saes[name] = sae

    def get_sae(self, name: str) -> Optional[SAEWrapper]:
        """Get a registered SAE.

        Args:
            name: SAE name

        Returns:
            SAE wrapper or None if not found
        """
        return self._saes.get(name)

    def analyze_activations(
        self,
        activations: Dict[str, np.ndarray],
    ) -> Dict[str, SAEFeatures]:
        """Analyze activations using registered SAEs.

        Args:
            activations: Dict mapping names to activation arrays

        Returns:
            Dict mapping names to SAEFeatures
        """
        results = {}

        for name, sae in self._saes.items():
            if name in activations:
                results[name] = sae.encode(activations[name])

        return results

    def find_behavior_features(
        self,
        clean_features: SAEFeatures,
        corrupt_features: SAEFeatures,
        threshold: float = 0.1,
    ) -> Dict[str, List[int]]:
        """Find features that differ between clean and corrupt traces.

        Args:
            clean_features: Features from clean run
            corrupt_features: Features from corrupted run
            threshold: Minimum difference to consider significant

        Returns:
            Dict with 'activated' and 'deactivated' feature lists
        """
        clean_arr = clean_features.activations.reshape(-1, clean_features.d_sae)
        corrupt_arr = corrupt_features.activations.reshape(-1, corrupt_features.d_sae)

        # Compare mean activation per feature
        clean_mean = np.mean(clean_arr, axis=0)
        corrupt_mean = np.mean(corrupt_arr, axis=0)
        diff = clean_mean - corrupt_mean

        activated = np.where(diff > threshold)[0].tolist()
        deactivated = np.where(diff < -threshold)[0].tolist()

        return {
            "activated": activated,  # More active in clean
            "deactivated": deactivated,  # Less active in clean
        }

    def feature_coactivation(
        self,
        features: SAEFeatures,
        top_k: int = 10,
    ) -> np.ndarray:
        """Compute feature co-activation matrix.

        Args:
            features: SAE features
            top_k: Consider only top-k features per position

        Returns:
            Co-activation matrix [d_sae, d_sae]
        """
        d_sae = features.d_sae
        coact = np.zeros((d_sae, d_sae), dtype=np.float32)

        top_features = features.top_k_features(top_k)

        for position_features in top_features:
            indices = [idx for idx, _ in position_features]
            for i in indices:
                for j in indices:
                    coact[i, j] += 1

        # Normalize
        num_positions = len(top_features)
        if num_positions > 0:
            coact /= num_positions

        return coact

File: microscope/src/sae.rs
//! Sparse Autoencoder Module
//!
//! This module provides tools for working with Sparse Autoencoders (SAEs):
//!
//! - SAE encoding and decoding
//! - Feature activation analysis
//! - Sparsity computation
//! - Top-k feature selection

use ndarray::{Array1, Array2, ArrayView1, ArrayView2, Axis};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Configuration for a Sparse Autoencoder
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SAEConfig {
    /// Input dimension (hidden size of the model)
    pub d_in: usize,
    /// Feature dimension (number of SAE features)
    pub d_sae: usize,
    /// Activation function type
    pub activation: ActivationType,
    /// Whether the encoder has bias
    pub encoder_bias: bool,
    /// Whether the decoder has bias
    pub decoder_bias: bool,
}

/// Activation function type for SAE
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ActivationType {
    /// ReLU activation
    ReLU,
    /// TopK activation (only keep top k activations)
    TopK(usize),
    /// JumpReLU (ReLU with threshold)
    JumpReLU,
}

/// Represents the weights of a trained SAE
#[derive(Debug, Clone)]
pub struct SAEWeights {
    /// Encoder weights [d_in, d_sae]
    pub w_enc: Array2<f32>,
    /// Decoder weights [d_sae, d_in]
    pub w_dec: Array2<f32>,
    /// Encoder bias [d_sae] (optional)
    pub b_enc: Option<Array1<f32>>,
    /// Decoder bias [d_in] (optional)
    pub b_dec: Option<Array1<f32>>,
    /// Configuration
    pub config: SAEConfig,
}

impl SAEWeights {
    /// Create new SAE weights
    pub fn new(
        w_enc: Array2<f32>,
        w_dec: Array2<f32>,
        b_enc: Option<Array1<f32>>,
        b_dec: Option<Array1<f32>>,
        config: SAEConfig,
    ) -> Self {
        Self {
            w_enc,
            w_dec,
            b_enc,
            b_dec,
            config,
        }
    }

    /// Get input dimension
    pub fn d_in(&self) -> usize {
        self.config.d_in
    }

    /// Get SAE feature dimension
    pub fn d_sae(&self) -> usize {
        self.config.d_sae
    }
}

/// Result of SAE encoding
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SAEFeatures {
    /// Feature activations [batch, seq_len, d_sae] or [batch * seq_len, d_sae]
    shape: Vec<usize>,
    data: Vec<f32>,
    /// Sparsity ratio (fraction of zero activations)
    pub sparsity: f32,
    /// Number of active features per position (mean)
    pub mean_active_features: f32,
}

impl SAEFeatures {
    /// Create from raw data
    pub fn new(shape: Vec<usize>, data: Vec<f32>) -> Self {
        let total = data.len() as f32;
        let zeros = data.iter().filter(|&&x| x == 0.0).count() as f32;
        let sparsity = zeros / total;

        // Compute mean active features per position
        let d_sae = *shape.last().unwrap_or(&1);
        let num_positions = data.len() / d_sae;
        let active_count: usize = data.iter().filter(|&&x| x > 0.0).count();
        let mean_active_features = active_count as f32 / num_positions as f32;

        Self {
            shape,
            data,
            sparsity,
            mean_active_features,
        }
    }

    /// Get as 2D array [positions, d_sae]
    /// Returns a zero array if shape is invalid (should not happen in normal use)
    pub fn as_array(&self) -> Array2<f32> {
        let d_sae = *self.shape.last().unwrap_or(&1);
        let positions: usize = if d_sae > 0 { self.data.len() / d_sae } else { 0 };
        Array2::from_shape_vec((positions, d_sae), self.data.clone())
            .unwrap_or_else(|_| Array2::zeros((1, 1)))
    }

    /// Get shape
    pub fn shape(&self) -> &[usize] {
        &self.shape
    }

    /// Get raw data
    pub fn data(&self) -> &[f32] {
        &self.data
    }

    /// Get indices of active features (above threshold)
    pub fn active_features(&self, threshold: f32) -> Vec<Vec<usize>> {
        let arr = self.as_array();
        arr.axis_iter(Axis(0))
            .map(|row| {
                row.iter()
                    .enumerate()
                    .filter(|(_, &v)| v > threshold)
                    .map(|(i, _)| i)
                    .collect()
            })
            .collect()
    }

    /// Get top-k feature indices and values per position
    pub fn top_k_features(&self, k: usize) -> Vec<Vec<(usize, f32)>> {
        let arr = self.as_array();
        arr.axis_iter(Axis(0))
            .map(|row| {
                let mut indexed: Vec<(usize, f32)> = row
                    .iter()
                    .enumerate()
                    .map(|(i, &v)| (i, v))
                    .collect();
                indexed.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
                indexed.truncate(k);
                indexed
            })
            .collect()
    }

    /// Compute feature frequency across all positions
    pub fn feature_frequency(&self, threshold: f32) -> Array1<f32> {
        let arr = self.as_array();
        let num_positions = arr.shape()[0] as f32;
        arr.map_axis(Axis(0), |col| {
            col.iter().filter(|&&v| v > threshold).count() as f32 / num_positions
        })
    }
}

/// SAE Encoder/Decoder operations
pub struct SAEEncoder;

impl SAEEncoder {
    /// Encode activations to SAE features
    pub fn encode(
        activations: ArrayView2<f32>,
        weights: &SAEWeights,
    ) -> SAEFeatures {
        // Pre-encoder: subtract decoder bias if present
        let centered = if let Some(ref b_dec) = weights.b_dec {
            let b_dec_2d = b_dec.view().insert_axis(Axis(0));
            &activations - &b_dec_2d
        } else {
            activations.to_owned()
        };

        // Encode: x @ W_enc + b_enc
        let mut features = centered.dot(&weights.w_enc);

        if let Some(ref b_enc) = weights.b_enc {
            features = &features + &b_enc.view().insert_axis(Axis(0));
        }

        // Apply activation
        let activated = match weights.config.activation {
            ActivationType::ReLU => {
                features.mapv(|x| if x > 0.0 { x } else { 0.0 })
            }
            ActivationType::TopK(k) => {
                Self::apply_topk(features.view(), k)
            }
            ActivationType::JumpReLU => {
                // JumpReLU with default threshold of 0.0
                features.mapv(|x| if x > 0.0 { x } else { 0.0 })
            }
        };

        let shape = vec![activated.shape()[0], activated.shape()[1]];
        SAEFeatures::new(shape, activated.into_raw_vec())
    }

    /// Apply top-k activation
    fn apply_topk(features: ArrayView2<f32>, k: usize) -> Array2<f32> {
        let mut result = Array2::zeros(features.raw_dim());

        for (i, row) in features.axis_iter(Axis(0)).enumerate() {
            let mut indexed: Vec<(usize, f32)> = row
                .iter()
                .enumerate()
                .map(|(j, &v)| (j, v))
                .collect();
            indexed.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));

            for (idx, val) in indexed.iter().take(k) {
                if *val > 0.0 {
                    result[[i, *idx]] = *val;
                }
            }
        }

        result
    }

    /// Decode SAE features back to activations
    pub fn decode(
        features: &SAEFeatures,
        weights: &SAEWeights,
    ) -> Array2<f32> {
        let features_arr = features.as_array();

        // Decode: features @ W_dec + b_dec
        let mut reconstructed = features_arr.dot(&weights.w_dec);

        if let Some(ref b_dec) = weights.b_dec {
            reconstructed = &reconstructed + &b_dec.view().insert_axis(Axis(0));
        }

        reconstructed
    }

    /// Compute reconstruction error
    pub fn reconstruction_error(
        original: ArrayView2<f32>,
        features: &SAEFeatures,
        weights: &SAEWeights,
    ) -> f32 {
        let reconstructed = Self::decode(features, weights);
        let diff = &original.to_owned() - &reconstructed;
        let mse = diff.mapv(|x| x * x).mean().unwrap_or(0.0);
        mse
    }
}

/// Feature analysis utilities
pub struct FeatureAnalyzer;

impl FeatureAnalyzer {
    /// Find features that activate for specific token patterns
    pub fn find_pattern_features(
        features_by_position: &[SAEFeatures],
        pattern_mask: &[bool],
        threshold: f32,
    ) -> HashMap<usize, f32> {
        let mut feature_activations: HashMap<usize, Vec<f32>> = HashMap::new();

        for (i, (features, &in_pattern)) in features_by_position.iter().zip(pattern_mask).enumerate() {
            if !in_pattern {
                continue;
            }

            let arr = features.as_array();
            // Assuming single position per SAEFeatures for this function
            if arr.shape()[0] > 0 {
                for (feat_idx, &val) in arr.row(0).iter().enumerate() {
                    if val > threshold {
                        feature_activations.entry(feat_idx).or_default().push(val);
                    }
                }
            }
        }

        // Compute mean activation per feature
        feature_activations
            .into_iter()
            .map(|(idx, vals)| {
                let mean = vals.iter().sum::<f32>() / vals.len() as f32;
                (idx, mean)
            })
            .collect()
    }

    /// Compute feature co-activation matrix
    pub fn feature_coactivation(
        features: &SAEFeatures,
        top_k: usize,
    ) -> Array2<f32> {
        let arr = features.as_array();
        let d_sae = arr.shape()[1];

        // Get top-k features per position
        let active_sets = features.top_k_features(top_k);

        // Build co-activation counts
        let mut coact = Array2::zeros((d_sae, d_sae));

        for active_features in &active_sets {
            for (i, _) in active_features {
                for (j, _) in active_features {
                    coact[[*i, *j]] += 1.0;
                }
            }
        }

        // Normalize by number of positions
        let num_positions = arr.shape()[0] as f32;
        coact /= num_positions;

        coact
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::Array2;

    fn create_test_weights() -> SAEWeights {
        let d_in = 64;
        let d_sae = 256;

        let config = SAEConfig {
            d_in,
            d_sae,
            activation: ActivationType::ReLU,
            encoder_bias: true,
            decoder_bias: true,
        };

        // Initialize random-ish weights
        let w_enc = Array2::from_shape_fn((d_in, d_sae), |(i, j)| {
            ((i + j) as f32 * 0.01).sin()
        });
        let w_dec = Array2::from_shape_fn((d_sae, d_in), |(i, j)| {
            ((i + j) as f32 * 0.01).cos()
        });
        let b_enc = Some(Array1::zeros(d_sae));
        let b_dec = Some(Array1::zeros(d_in));

        SAEWeights::new(w_enc, w_dec, b_enc, b_dec, config)
    }

    #[test]
    fn test_sae_encode_decode() {
        let weights = create_test_weights();
        let activations = Array2::from_shape_fn((4, 64), |(i, j)| {
            ((i * j) as f32 * 0.1).tanh()
        });

        let features = SAEEncoder::encode(activations.view(), &weights);

        assert_eq!(features.shape(), &[4, 256]);
        assert!(features.sparsity >= 0.0 && features.sparsity <= 1.0);
    }

    #[test]
    fn test_sae_features_active() {
        let weights = create_test_weights();
        let activations = Array2::ones((2, 64));

        let features = SAEEncoder::encode(activations.view(), &weights);
        let active = features.active_features(0.0);

        assert_eq!(active.len(), 2);  // 2 positions
    }

    #[test]
    fn test_sae_topk_features() {
        let weights = create_test_weights();
        let activations = Array2::ones((2, 64));

        let features = SAEEncoder::encode(activations.view(), &weights);
        let top5 = features.top_k_features(5);

        assert_eq!(top5.len(), 2);
        assert!(top5[0].len() <= 5);
    }

    #[test]
    fn test_reconstruction_error() {
        let weights = create_test_weights();
        let activations = Array2::ones((2, 64));

        let features = SAEEncoder::encode(activations.view(), &weights);
        let error = SAEEncoder::reconstruction_error(activations.view(), &features, &weights);

        // Error should be non-negative
        assert!(error >= 0.0);
    }

    #[test]
    fn test_feature_frequency() {
        let features = SAEFeatures::new(
            vec![4, 10],
            vec![
                1.0, 0.0, 0.5, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                0.0, 0.0, 0.5, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            ],
        );

        let freq = features.feature_frequency(0.0);

        // Feature 0 is active in 3/4 positions
        assert!((freq[0] - 0.75).abs() < 0.01);
        // Feature 1 is active in 1/4 positions
        assert!((freq[1] - 0.25).abs() < 0.01);
    }
}
